src/main/java/de/tum/in/www1/artemis/service/notifications/GeneralInstantNotificationService.java

```diff
@@ -0,0 +1,90 @@
+        this.applePushNotificationService = applePushNotificationService;
+        this.firebasePushNotificationService = firebasePushNotificationService;
+        this.mailService = mailService;
+        this.notificationSettingsService = notificationSettingsService;
+    }
+
+    /**
+     * Checks for the user if the notification should be sent as an email or/and as a push notification.
+     * Then delegates the actual sending to the corresponding {@link InstantNotificationService}s
+     */
+    @Override
+    public void sendNotification(Notification notification, User user, Object notificationSubject) {
+        boolean allowsEmail = notificationSettingsService.checkIfNotificationIsAllowedInCommunicationChannelBySettingsForGivenUser(notification, user, EMAIL);
+
+        boolean allowsPush = notificationSettingsService.checkIfNotificationIsAllowedInCommunicationChannelBySettingsForGivenUser(notification, user, PUSH);
+
+        NotificationType type = NotificationConstants.findCorrespondingNotificationType(notification.getTitle());
+        if (allowsEmail && notificationSettingsService.checkNotificationTypeForEmailSupport(type)) {
+            mailService.sendNotification(notification, user, notificationSubject);
+        }
+
+        if (allowsPush) {
+            applePushNotificationService.sendNotification(notification, user, notificationSubject);
+            firebasePushNotificationService.sendNotification(notification, user, notificationSubject);
+        }
+    }
+
+    /**
+     * Checks for each user if the notification should be sent as an email or/and as a push notification.
+     * Then delegates the actual sending to the corresponding {@link InstantNotificationService}s
+     */
+    @Async
+    @Override
+    public void sendNotification(Notification notification, List<User> users, Object notificationSubject) {
+        SecurityUtils.setAuthorizationObject();
+
+        var emailRecipients = filterRecipients(notification, users, EMAIL);
+        var pushRecipients = filterRecipients(notification, users, PUSH);
+
+        applePushNotificationService.sendNotification(notification, pushRecipients, notificationSubject);
+        firebasePushNotificationService.sendNotification(notification, pushRecipients, notificationSubject);
+
+        NotificationType type = NotificationConstants.findCorrespondingNotificationType(notification.getTitle());
+        if (notificationSettingsService.checkNotificationTypeForEmailSupport(type)) {
+            mailService.sendNotification(notification, emailRecipients, notificationSubject);
+        }
+    }
+
+    @NotNull
+    private List<User> filterRecipients(Notification notification, List<User> users, NotificationSettingsCommunicationChannel channel) {
+        return users.stream().filter(user -> notificationSettingsService.checkIfNotificationIsAllowedInCommunicationChannelBySettingsForGivenUser(notification, user, channel))
```

Comment:
The filter executes at least one database query for each user. Can you filter them in a single query?